Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/TargetController.cs

```diff
@@ -0,0 +1,143 @@
+using Unity.MLAgents;
+using UnityEngine.Events;
+
+namespace Unity.MLAgentsExamples
+{
+    /// <summary>
+    /// Utility class to allow target placement and collision detection with an agent
+    /// Add this script to the target you want the agent to touch.
+    /// Callbacks will be triggered any time the target is touched with a collider tagged as 'tagToDetect'
+    /// </summary>
+    public class TargetController : MonoBehaviour
+    {
+        
+        [Header("Collider Tag To Detect")]
+        public string tagToDetect = "agent"; //collider tag to detect 
+        
+        [Header("Target Placement")]
+        public float spawnRadius; //The radius in which a target can be randomly spawned.
+        public bool respawnIfTouched; //Should the target respawn to a different position when touched
+        
+        [Header("Target Fell Protection")]
+        public bool respawnIfFallsOffPlatform = true; //If the target falls off the platform, reset the position.
+        public float fallDistance = 5; //distance below the starting height that will trigger a respawn 
+
+
+        private Vector3 m_startingPos; //the starting position of the target
+        private Agent m_agentTouching; //the agent currently touching the target
+
+        [System.Serializable]
+        public class TriggerEvent : UnityEvent<Collider>
+        {
+        }
+
+        [Header("Trigger Callbacks")] public bool triggerIsTouching;
+        public TriggerEvent onTriggerEnterEvent = new TriggerEvent();
+        public TriggerEvent onTriggerStayEvent = new TriggerEvent();
+        public TriggerEvent onTriggerExitEvent = new TriggerEvent();
+
+        [System.Serializable]
+        public class CollisionEvent : UnityEvent<Collision>
+        {
+        }
+
+        [Header("Collision Callbacks")] public bool colliderIsTouching;
+        public CollisionEvent onCollisionEnterEvent = new CollisionEvent();
+        public CollisionEvent onCollisionStayEvent = new CollisionEvent();
+        public CollisionEvent onCollisionExitEvent = new CollisionEvent();
+
+        // Start is called before the first frame update
+        void OnEnable()
+        {
+            m_startingPos = transform.position;
+            if (respawnIfTouched)
+            {
+                MoveTargetToRandomPosition();
+            }
+        }
+
+        void Update()
+        {
+            if (respawnIfFallsOffPlatform)
+            {
+                if (transform.position.y < m_startingPos.y - fallDistance)
+                {
+                    Debug.LogError($"{transform.name} Fell Off Platform");
+                    MoveTargetToRandomPosition();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Moves target to a random position within specified radius.
+        /// </summary>
+        public void MoveTargetToRandomPosition()
+        {
+            var newTargetPos = m_startingPos + (Random.insideUnitSphere * spawnRadius);
```

Comment:
Is it a 3D sphere or a 2D sphere ? I think we want it to be 2D. Or it might spawn at `m_startingPos - spawnRadius * Vector3.up`

Reply:
i this you're right i'll doublecheck.

Reply:
That is still not a uniform circle, probability of spawning at the center of the circle will be a lot higher than the probability of spawning at the edges (because it is a squished sphere). But I guess it does not matter...

Reply:
oh ok i understand what you are saying now. yes you are correct we are doing a random point in a sphere then squishing the Y. the reason i didn't use Random.InsideUnitCircle is because a Random.InsideUnitCircle returns a Vector2 so then you have to "remap" that Vector2 to a Vector3 and then set the Y position. it adds one more line of code and yields a very similar result.